File: CryptArithmetic.java
import java.util.*;

class CryptArithmetic{
    static boolean solved = false;
    public static void main(String args[]){
        Scanner in = new Scanner(System.in);
        String word1 = "", word2 = "", result = "";
        System.out.println("\nCRYPTARITHMETIC PUZZLE SOLVER");
        System.out.println("WORD1 + WORD2 = RESULT");
        System.out.print("Enter WORD1: ");
        word1 = in.next().toUpperCase();
        System.out.print("Enter WORD2: ");
        word2 = in.next().toUpperCase();
        System.out.print("Enter RESULT: ");
        result = in.next().toUpperCase();
        if(!word1.matches("^[A-Z]*$") || !word2.matches("^[A-Z]*$") || !result.matches("^[A-Z]*$"))
            System.out.println("\nOnly Letters allowed.");
        else if(result.length() > Math.max(word1.length(), word2.length()) + 1)
            System.out.println("\n0 Solutions!");
        else{
            Set<Character> set = new HashSet<Character>();
            for(char c:word1.toCharArray())
                set.add(c);
            for(char c:word2.toCharArray())
                set.add(c);
            for(char c:result.toCharArray())
                set.add(c);
            if(set.size() > 10){
                System.out.println("\n0 solutions!");
                System.exit(0);
            }
            List<Character> letters = new ArrayList<Character>(set);
            System.out.println("\nSolutions:");
            solve(letters, new ArrayList<Integer>(), new boolean[10], new String[]{word1, word2, result});
            if(!solved)
                System.out.println("\n0 solutions!");
        }
        in.close();
    }
    static void solve(List<Character> letters, List<Integer> values, boolean[] visited, String[] words){
        if(letters.size() == values.size()){
            TreeMap<Character,Integer> map = new TreeMap<Character,Integer>();
            for(int i=0;i<letters.size();i++)
                map.put(letters.get(i), values.get(i));
            if(map.get(words[0].charAt(0)) == 0 || map.get(words[1].charAt(0)) == 0 || map.get(words[2].charAt(0)) == 0)
                return;
            String word1 = "", word2 = "", res = "";
            for(char c: words[0].toCharArray())
                word1 += map.get(c);
            for(char c: words[1].toCharArray())
                word2 += map.get(c);
            for(char c: words[2].toCharArray())
                res += map.get(c);
            if((Integer.parseInt(word1) + Integer.parseInt(word2)) == Integer.parseInt(res)){
                System.out.println(word1 + " + " + word2 + " = " + res+"\t"+map);
                solved = true;
            }
        }
        for(int i=0;i<10;i++){
            if(!visited[i]){
                visited[i] = true;
                values.add(i);
                solve(letters, values, visited, words);
                values.remove(values.size() - 1);
                visited[i] = false;
            }
        }
    }
}

File: CryptArithmetic.py
'''Cryptarithmetic puzzles are arithmetic problems made of letters instead of
numbers. The goal is to replace each letter with a single, unique digit in order
to make the arithmetic work out correctly.

Example:  S E N D + M O R E = M O N E Y

   S E N D
+  M O R E
----------
 M O N E Y
----------
Vales which satisfies this constraints
S = 9, E = 5, N = 6, D = 7, M = 1, O = 0, R = 8, Y = 2
'''


def find_value(word, assigned):
  num = 0
  for char in word:
    num = num * 10
    num += assigned[char]
  return num


def is_valid_assignment(word1, word2, result, assigned):
  # First letter of any word cannot be zero.
  if assigned[word1[0]] == 0 or assigned[word2[0]] == 0 or assigned[
      result[0]] == 0:
    return False
  return True


def _solve(word1, word2, result, letters, assigned, solutions):
  if not letters:
    if is_valid_assignment(word1, word2, result, assigned):
      num1 = find_value(word1, assigned)
      num2 = find_value(word2, assigned)
      num_result = find_value(result, assigned)
      if num1 + num2 == num_result:
        solutions.append((f'{num1} + {num2} = {num_result}', assigned.copy()))
    return

  for num in range(10):
    if num not in assigned.values():
      cur_letter = letters.pop()
      assigned[cur_letter] = num
      _solve(word1, word2, result, letters, assigned, solutions)
      assigned.pop(cur_letter)
      letters.append(cur_letter)


def solve(word1, word2, result):
  letters = sorted(set(word1) | set(word2) | set(result))
  if len(result) > max(len(word1), len(word2)) + 1 or len(letters) > 10:
    print('0 Solutions!')
    return

  solutions = []
  _solve(word1, word2, result, letters, {}, solutions)
  if solutions:
    print('\nSolutions:')
    for soln in solutions:
      print(f'{soln[0]}\t{soln[1]}')


if __name__ == '__main__':
  print('CRYPTARITHMETIC PUZZLE SOLVER')
  print('WORD1 + WORD2 = RESULT')
  word1 = input('Enter WORD1: ').upper()
  word2 = input('Enter WORD2: ').upper()
  result = input('Enter RESULT: ').upper()

  if not word1.isalpha() or not word2.isalpha() or not result.isalpha():
    raise TypeError('Inputs should ony consists of alphabets.')

  solve(word1, word2, result)
